File: backend/app/schemas/codeforces.py
"""
Codeforces schemas.

Unlike app/schemas/leetcode.py, there's no pre-existing frontend
contract constraining these shapes — the Codeforces frontend is being
built in this same module. Schemas are designed around Codeforces's
genuinely richer API (real verdicts, ratings, contest history)
rather than mirroring LeetCode's narrower one.
"""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, computed_field


class CodeforcesProfileResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    user_id: int
    handle: str
    avatar_url: str | None
    country: str | None
    organization: str | None
    rank: str | None
    rating: int | None
    max_rank: str | None
    max_rating: int | None
    contribution: int
    total_solved: int
    synced_at: datetime | None


class CodeforcesSubmissionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    submission_id: int = Field(validation_alias="codeforces_submission_id")
    contest_id: int | None
    problem_index: str
    problem_name: str
    problem_rating: int | None
    programming_language: str
    verdict: str | None
    passed_test_count: int
    time_consumed_millis: int
    memory_consumed_bytes: int
    creation_time_seconds: int


class CodeforcesContestResultResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    contest_id: int
    contest_name: str
    rank: int
    old_rating: int
    new_rating: int
    rating_update_time_seconds: int

    @computed_field  # type: ignore[prop-decorator]
    @property
    def rating_change(self) -> int:
        return self.new_rating - self.old_rating


class CodeforcesSyncResponse(BaseModel):
    message: str
    synced_at: str
    submissions_saved: int
    contests_saved: int


class CodeforcesConnectRequest(BaseModel):
    handle: str = Field(min_length=1, max_length=100)

File: frontend/src/types/codeforces.ts
/**
 * Codeforces types.
 *
 * Mirrors backend/app/schemas/codeforces.py. Field names are
 * snake_case to match the FastAPI JSON output directly — no
 * camelCase transformation, consistent with how we handle
 * LeetCode and auth types throughout the app.
 */

export interface CodeforcesProfile {
  id: number;
  user_id: number;
  handle: string;
  avatar_url: string | null;
  country: string | null;
  organization: string | null;
  rank: string | null;
  rating: number | null;
  max_rank: string | null;
  max_rating: number | null;
  contribution: number;
  total_solved: number;
  synced_at: string | null;
}

export interface CodeforcesSubmission {
  submission_id: number;
  contest_id: number | null;
  problem_index: string;
  problem_name: string;
  problem_rating: number | null;
  programming_language: string;
  verdict: string | null;
  passed_test_count: number;
  time_consumed_millis: number;
  memory_consumed_bytes: number;
  creation_time_seconds: number;
}

export interface CodeforcesContestResult {
  contest_id: number;
  contest_name: string;
  rank: number;
  old_rating: number;
  new_rating: number;
  rating_update_time_seconds: number;
  rating_change: number;
}

export interface CodeforcesSyncResponse {
  message: string;
  synced_at: string;
  submissions_saved: number;
  contests_saved: number;
}

export interface CodeforcesConnectRequest {
  handle: string;
}

/** Codeforces rank colour thresholds — matches the official CF rank system. */
export const CF_RANK_COLORS: Record<string, string> = {
  newbie: "#808080",
  pupil: "#008000",
  specialist: "#03a89e",
  expert: "#0000ff",
  "candidate master": "#aa00aa",
  master: "#ff8c00",
  "international master": "#ff8c00",
  grandmaster: "#ff0000",
  "international grandmaster": "#ff0000",
  "legendary grandmaster": "#ff0000",
};
